smartconfig/file.py

```diff
@@ -0,0 +1,94 @@
+    """
+    Recursively update the `dest` mapping with values from `source`.
+
+    Overwrite any non-mapping value in `dest` with the value from `source`.
+    Expand keys which are strings of dot-delimited paths into nested mapping nodes. For example,
+
+    a.b.c: 1
+
+    becomes
+
+    a:
+      b:
+        c: 1
+
+    Args:
+        source: The mapping with the new values.
+        dest: The mapping to update.
+        path: The dot-delimited path to `dest`.
+
+    Returns:
+        The updated dest mapping.
+    """
+    for key, value in source.items():
+        if '.' in key:
+            key, child_node = key.split('.', maxsplit=1)
+            dest[key] = _update_mapping({child_node: value}, dest.get(key, {}), path + key)
+
+        elif isinstance(value, Mapping):
+            dest[key] = _update_mapping(value, dest.get(key, {}), path + key)
+        else:
+            _registry.overwritten_attributes.add(path + key)
+            dest[key] = value
+
+    return dest
+
+
+def load(path: Union[str, bytes, PathLike]) -> None:
+    """
+    Read a YAML file at `path` and update the configuration with its values.
+
+    Overwrite default values set in `ConfigEntry` objects with values from the YAML file. A YAML node overwrites an
+    attribute of a `ConfigEntry` if their paths are equal. The path of the node is the dot-delimited concatenation of
+    its parents' keys with its own key. In the following example, the path of the node 'class' is 'module.class'.
+
+    module:
+      class:
+        attribute_1: value
+
+    Nodes may be fully expanded (like above), partially collapsed, or fully collapsed. A `ConfigEntry` node's definition
+    could be split by using different levels of expansion. For example,
+
+    module:
+      class:
+        attribute_1: value
+
+    module.class:
+      attribute_2: value
+
+    module.class:
+      attribute_3: value
+
+    would overwrite `attribute_1` and `attribute_3` of the `ConfigEntry` for the path 'package.module'. When a key is
+    defined multiple times, the last definition is the one used. Therefore, `attribute_2` is not included.
+
+    The YAML does not strictly have to contain nodes that correspond to `ConfigEntry` objects. For example, it's valid
+    to have a non-`ConfigEntry` node marked by an anchor which is later referenced in some node that does correspond to
+    a `ConfigEntry`. There could be nodes that are simply not used at all. It's even valid for the root to not be a
+    mapping node. Such file would effectively configure nothing, but loading it is still supported.
+
+    Args:
+        path: The path to the configuration file.
+
+    Raises:
+        FileNotFoundError: The configuration file doesn't exist.
+        IOError: An error occurred when reading the file.
+        yaml.YAMLError: PyYAML failed to load the YAML.
+    """
+    with open(path) as file:
+        yaml_content = yaml.full_load(file)
```

Comment:
While this is convenient, it is inflexible. Users cannot customise how the YAML is loaded. They also cannot customise the encoding used to open the file. I think we can just add the loader and encoding as parameters to our `load` function without losing too much flexibility; `yaml.load` doesn't take any other arguments besides the loader and `open`'s remaining arguments are niche. That being said, it may be worthwhile to add a `load_stream` function in case the YAML is not a local file.

Reply:
I switched to an isolated YAML loader in 1ce3179, with the `!REF` construct already in. I also added a `loader` keyword argument to the `load()` function to select which loader you want to use. I think we can assume that the user doesn't want the `!REF` constructor if they use another loader, or will have to add it manually.

I didn't add any `load_stream()` function since this can be done through a `StringIO`, do you think that's really needed?

Reply:
> I didn't add any load_stream() function since this can be done through a StringIO, do you think that's really needed?

Yes, it would be useful because `StringIO` provides a file-like object while the `load()` function currently expects a _path_.

Furthermore, I still think an `encoding` parameter to forward to `open()` would be useful to have on `load()`.

Reply:
Resolved by 454ad1c and 2ecf31b